File: models/Issue.js
const mongoose = require('mongoose');

const MediaSchema = new mongoose.Schema({
    url: { type: String, required: true },
    publicId: { type: String },
    type: { type: String, enum: ['image', 'video'], required: true }
}, { _id: false });

const TimelineEntrySchema = new mongoose.Schema({
    stage: {
        type: String,
        enum: ['review', 'started', 'inprogress', 'completed'],
        required: true
    },
    note: { type: String, maxlength: 1000 },
    media: [MediaSchema],
    updatedBy: { type: mongoose.Schema.ObjectId, ref: 'User', required: true },
    updatedAt: { type: Date, default: Date.now }
}, { _id: true });

const IssueSchema = new mongoose.Schema({
    title: {
        type: String,
        required: [true, 'Please provide an issue title'],
        trim: true,
        maxlength: [200, 'Title cannot exceed 200 characters']
    },
    description: {
        type: String,
        required: [true, 'Please provide a description'],
        maxlength: [2000, 'Description cannot exceed 2000 characters']
    },
    category: {
        type: String,
        enum: ['electrical', 'plumbing', 'structural', 'water', 'security', 'cleaning', 'internet', 'other'],
        default: 'other'
    },
    priority: {
        type: String,
        enum: ['low', 'medium', 'high', 'urgent'],
        default: 'medium'
    },
    status: {
        type: String,
        enum: ['review', 'started', 'inprogress', 'completed', 'cancelled'],
        default: 'review'
    },
    reporter: {
        type: mongoose.Schema.ObjectId,
        ref: 'User',
        required: true
    },
    estate: {
        type: mongoose.Schema.ObjectId,
        ref: 'Estate'
    },
    unit: {
        type: mongoose.Schema.ObjectId,
        ref: 'Unit'
    },
    tenant: {
        type: mongoose.Schema.ObjectId,
        ref: 'Tenant'
    },
    assignedTo: {
        type: mongoose.Schema.ObjectId,
        ref: 'User'
    },
    // Initial media (images/videos attached at report time)
    media: [MediaSchema],
    // Audit trail: every stage change with optional proof media
    timeline: [TimelineEntrySchema],
    resolvedAt: { type: Date },
    isActive: { type: Boolean, default: true }
}, { timestamps: true });

IssueSchema.index({ reporter: 1, status: 1 });
IssueSchema.index({ estate: 1, status: 1 });
IssueSchema.index({ assignedTo: 1, status: 1 });
IssueSchema.index({ createdAt: -1 });

module.exports = mongoose.model('Issue', IssueSchema);

File: fastapi_app/models/issue.py
from sqlalchemy import String, Boolean, DateTime, JSON, Text
from sqlalchemy.orm import Mapped, mapped_column
from models.base import Base, gen_uuid
from datetime import datetime
from utils.time_utils import utcnow


class Issue(Base):
    __tablename__ = "issues"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=gen_uuid)
    title: Mapped[str] = mapped_column(String(255), default="")
    description: Mapped[str] = mapped_column(Text, default="")
    category: Mapped[str] = mapped_column(String(100), default="other")
    priority: Mapped[str] = mapped_column(String(50), default="medium")
    status: Mapped[str] = mapped_column(String(50), default="open")
    stage: Mapped[str] = mapped_column(String(50), default="review")
    reporter: Mapped[str | None] = mapped_column(String(36), nullable=True)
    assigned_to: Mapped[str | None] = mapped_column(String(36), nullable=True)
    estate: Mapped[str | None] = mapped_column(String(36), nullable=True, index=True)
    unit: Mapped[str | None] = mapped_column(String(36), nullable=True)
    tenant: Mapped[str | None] = mapped_column(String(36), nullable=True)
    media: Mapped[list] = mapped_column(JSON, default=list)
    timeline: Mapped[list] = mapped_column(JSON, default=list)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)
